xwiki-rendering-syntaxes/xwiki-rendering-syntax-xwiki20/src/main/java/org/xwiki/rendering/internal/parser/xwiki20/wikimodel/XWiki20XWikiGeneratorListener.java

```diff
@@ -0,0 +1,189 @@
+    {
+        flushFormat();
+
+        QueueListener queue = (QueueListener) getListener();
+        popListener();
+
+        Map<String, String> paragraphParameters = convertParameters(params);
+
+        // If the only content of the paragraph is an image with a label, potentially wrapped in a link, convert to a
+        // figure with the parsed label as caption.
+
+        QueueListener.Event imageEvent = null;
+
+        if (queue.size() == 1 && queue.getFirst().eventType == EventType.ON_IMAGE) {
+            imageEvent = queue.getFirst();
+        } else if (isLinkWrappedImage(queue)) {
+            imageEvent = queue.get(1);
+        }
+
+        if (imageEvent != null && imageEvent.eventParameters[3] instanceof Map<?, ?>
+            && ((Map<?, ?>) imageEvent.eventParameters[3]).containsKey(IMAGE_LABEL_PARAMETER))
+        {
+            Map<?, ?> imageParameters = (Map<?, ?>) imageEvent.eventParameters[3];
+            String label = (String) imageParameters.get(IMAGE_LABEL_PARAMETER);
+
+            // Remove the label from the image parameters.
+            if (imageParameters.size() == 1) {
+                imageEvent.eventParameters[3] = Listener.EMPTY_PARAMETERS;
+            } else {
+                imageParameters.remove(IMAGE_LABEL_PARAMETER);
+            }
+
+            Map<String, String> figureParameters = new LinkedHashMap<>(paragraphParameters);
+            figureParameters.merge(CLASS_PARAMETER, IMAGE_CLASS,
+                (oldValue, newValue) -> {
+                    if (Arrays.asList(StringUtils.split(oldValue)).contains(newValue)) {
+                        return oldValue;
+                    } else {
+                        return oldValue + " " + newValue;
+                    }
+                });
+
+            // If this should be changed to produce more than just an image directly inside the caption,
+            // CaptionedImageParseListenerProvider for xwiki-platform needs to be adapted.
+            getListener().beginFigure(figureParameters);
+            queue.consumeEvents(getListener());
+            getListener().beginFigureCaption(Listener.EMPTY_PARAMETERS);
+            try {
+                // Render the caption ignoring begin/endDocument events.
+                WrappingListener wrapper = new WrappingListener()
+                {
+                    @Override
+                    public void beginDocument(MetaData metadata)
+                    {
+                        // ignore.
+                    }
+
+                    @Override
+                    public void endDocument(MetaData metadata)
+                    {
+                        // ignore.
+                    }
+                };
+                wrapper.setWrappedListener(getListener());
+                this.imageLabelParser.parse(new StringReader(label), wrapper, getIdGenerator());
+            } catch (ParseException e) {
+                // TODO what should we do here ?
```

Comment:
Good question. This is the case where the label of an image link fails to be parsed right?

Reply:
Note that this code is not new, I've just moved it. The TODO was already there before, and we have the same TODO for link content which is where I got this from.

Ignoring the exception here basically means that the content will be lost during rendering if there is an exception. A better idea is probably to display an error block with exception information as we do in other places.

Reply:
ok, so an interesting improvement, but definitely a separate issue. Thanks